### ISSUE TITLE
GLtexture deinit after end of main

### ISSUE THREAD

**Issue by jengelh** [sirjuddington/SLADE#1183]
**SLADE Version**: 3.1.12
**Operating System**: Linux

**Issue Details**:
There is a MapTextureManager object with static storage duration and whose destruction happens after libGL has already been deinitialized; or this is a double free. This manifests as a crash, irregularly, when exiting SLADE. Specifically, this happens for the `MapTextureManager::editor_images` member.

```
#0  0x00007fa45fc85b74 in _int_free (av=0x7fa45ffb5aa0 <main_arena>, p=0x55d75c59d490, have_lock=0)
    at malloc.c:4323
#1  0x00007fa444b5f710 in _mesa_reference_texobj_ (ptr=0x7ffdc157c720, tex=0x0) at main/texobj.c:570
#2  0x00007fa444b5fadc in _mesa_reference_texobj (tex=0x0, ptr=0x7ffdc157c720) at main/texobj.h:94
#3  0x00007fa444b5fadc in delete_textures (ctx=0x55d75a179760, n=<optimized out>, textures=<optimized out>)
    at main/texobj.c:1508
#4  0x000055d7511f8021 in GLTexture::clear() (this=0x55d75ba969e0)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/OpenGL/GLTexture.cpp:372
#5  0x000055d7511f809d in GLTexture::~GLTexture() (this=0x55d75ba969e0, __in_chrg=<optimized out>)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/OpenGL/GLTexture.cpp:70
#6  0x000055d7510e7f6b in map_tex_t::~map_tex_t() (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/MapEditor/MapTextureManager.h:13
#7  0x000055d7510e8128 in std::pair<wxString const, map_tex_t>::~pair()
    (this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/7/bits/stl_pair.h:208
#8  0x000055d7510e8128 in __gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<wxString const, map_tex_t> > >::destroy<std::pair<wxString const, map_tex_t> >(std::pair<wxString const, map_tex_t>*)
    (this=<optimized out>, __p=<optimized out>) at /usr/include/c++/7/ext/new_allocator.h:140
#9  0x000055d7510e8128 in std::allocator_traits<std::allocator<std::_Rb_tree_node<std::pair<wxString const, map_tex_t> > > >::destroy<std::pair<wxString const, map_tex_t> >(std::allocator<std::_Rb_tree_node<std::pair<wxString const, map_tex_t> > >&, std::pair<wxString const, map_tex_t>*) (__a=..., __p=<optimized out>)
    at /usr/include/c++/7/bits/alloc_traits.h:487
#10 0x000055d7510e8128 in std::_Rb_tree<wxString, std::pair<wxString const, map_tex_t>, std::_Select1st<std::pair<wxString const, map_tex_t> >, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::_M_destroy_node(std::_Rb_tree_node<std::pair<wxString const, map_tex_t> >*)
    (this=0x55d751af8aa0 <MapEditor::texture_manager+192>, __p=0x55d75ba93610)
    at /usr/include/c++/7/bits/stl_tree.h:650
#11 0x000055d7510e8128 in std::_Rb_tree<wxString, std::pair<wxString const, map_tex_t>, std::_Select1st<std::pai--Type <RET> for more, q to quit, c to continue without paging--
r<wxString const, map_tex_t> >, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::_M_drop_node(std::_Rb_tree_node<std::pair<wxString const, map_tex_t> >*)
    (this=0x55d751af8aa0 <MapEditor::texture_manager+192>, __p=0x55d75ba93610)
    at /usr/include/c++/7/bits/stl_tree.h:658
#12 0x000055d7510e8128 in std::_Rb_tree<wxString, std::pair<wxString const, map_tex_t>, std::_Select1st<std::pair<wxString const, map_tex_t> >, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::_M_erase(std::_Rb_tree_node<std::pair<wxString const, map_tex_t> >*)
    (this=this@entry=0x55d751af8aa0 <MapEditor::texture_manager+192>, __x=0x55d75ba93610)
    at /usr/include/c++/7/bits/stl_tree.h:1858
#13 0x000055d7510e811b in std::_Rb_tree<wxString, std::pair<wxString const, map_tex_t>, std::_Select1st<std::pair<wxString const, map_tex_t> >, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::_M_erase(std::_Rb_tree_node<std::pair<wxString const, map_tex_t> >*)
    (this=this@entry=0x55d751af8aa0 <MapEditor::texture_manager+192>, __x=0x55d75c554290)
    at /usr/include/c++/7/bits/stl_tree.h:1856
#14 0x000055d7510dd6c0 in std::_Rb_tree<wxString, std::pair<wxString const, map_tex_t>, std::_Select1st<std::pair<wxString const, map_tex_t> >, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::~_Rb_tree() (this=0x55d751af8aa0 <MapEditor::texture_manager+192>, __in_chrg=<optimized out>)
    at /usr/include/c++/7/bits/stl_tree.h:949
#15 0x000055d7510dd6c0 in std::map<wxString, map_tex_t, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::~map() (this=0x55d751af8aa0 <MapEditor::texture_manager+192>, __in_chrg=<optimized out>)
    at /usr/include/c++/7/bits/stl_map.h:294
#16 0x000055d7510dd6c0 in MapTextureManager::~MapTextureManager()
    (this=0x55d751af89e0 <MapEditor::texture_manager>, __in_chrg=<optimized out>)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/MapEditor/MapTextureManager.cpp:71
#17 0x00007fa45fc3d138 in __run_exit_handlers
    (status=0, listp=0x7fa45ffb5578 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:83
#18 0x00007fa45fc3d18a in __GI_exit (status=<optimized out>) at exit.c:105
#19 0x00007fa45fc25351 in __libc_start_main (main=
    0x55d750d0d3f0 <main(int, char**)>, argc=2, argv=0x7ffdc157c948, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffdc157c938) at ../csu/libc-start.c:342
#20 0x000055d750d2b89a in _start () at ../sysdeps/x86_64/start.S:120

(gdb) 
#4  0x000055d7511f8021 in GLTexture::clear (this=0x55d75ba969e0)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/OpenGL/GLTexture.cpp:372
372                     glDeleteTextures(1, &tex[a].id);
(gdb) 
#5  0x000055d7511f809d in GLTexture::~GLTexture (this=0x55d75ba969e0, __in_chrg=<optimized out>)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/OpenGL/GLTexture.cpp:70
70                      clear();
(gdb) 
#6  0x000055d7510e7f6b in map_tex_t::~map_tex_t (this=<optimized out>, __in_chrg=<optimized out>)
    at /usr/src/debug/slade-3.1.12-0.x86_64/src/MapEditor/MapTextureManager.h:13
13              ~map_tex_t() { if (texture && texture != &(GLTexture::missingTex())) delete texture; }
(gdb) 
#7  0x000055d7510e8128 in std::pair<wxString const, map_tex_t>::~pair (this=<optimized out>, 
    __in_chrg=<optimized out>) at /usr/include/c++/7/bits/stl_pair.h:208
208         struct pair
(gdb) 
#15 std::map<wxString, map_tex_t, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > >::~map (this=0x55d751af8aa0 <MapEditor::texture_manager+192>, __in_chrg=<optimized out>)
    at /usr/include/c++/7/bits/stl_map.h:294
294           ~map() = default;
(gdb) 
#16 MapTextureManager::~MapTextureManager (this=0x55d751af89e0 <MapEditor::texture_manager>, 
    __in_chrg=<optimized out>) at /usr/src/debug/slade-3.1.12-0.x86_64/src/MapEditor/MapTextureManager.cpp:71
71      MapTextureManager::~MapTextureManager()

$2 = (std::map<wxString, map_tex_t, std::less<wxString>, std::allocator<std::pair<wxString const, map_tex_t> > > * const) 0x55d751af8aa0 <MapEditor::texture_manager+192>
(gdb) p *this
$3 = {_M_t = {
    _M_impl = {<std::allocator<std::_Rb_tree_node<std::pair<wxString const, map_tex_t> > >> = {<__gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<wxString const, map_tex_t> > >> = {<No data fields>}, <No data fields>}, <std::_Rb_tree_key_compare<std::less<wxString> >> = {
        _M_key_compare = {<std::binary_function<wxString, wxString, bool>> = {<No data fields>}, <No data fields>}}, <std::_Rb_tree_header> = {_M_header = {_M_color = std::_S_red, _M_parent = 0x55d75c5a7950, 
          _M_left = 0x55d75c52fac0, _M_right = 0x55d75c5e5c50}, _M_node_count = 100}, <No data fields>}}}

(so, 100 textures left)
```

**Comment by sirjuddington** [comment]
Does this still happen in 3.2.0? The gl texture handling is done completely differently there

**Comment by sirjuddington** [comment]
Going to close this one, unlikely to be fixed in 3.1.x but feel free to reopen if it's still a thing in 3.2.0
